src/lib/pubkey/ec_group/ec_inner_pc.cpp

```diff
@@ -0,0 +1,210 @@
+                                                           std::span<const uint8_t> bytes) {
+   if(auto pt = curve.deserialize_point(bytes)) {
+      return *pt;
+   } else {
+      throw Decoding_Error("Invalid elliptic curve point encoding");
+   }
+}
+
+}  // namespace
+
+const EC_Scalar_Data_PC& EC_Scalar_Data_PC::checked_ref(const EC_Scalar_Data& data) {
+   const auto* p = dynamic_cast<const EC_Scalar_Data_PC*>(&data);
+   if(!p) {
+      throw Invalid_State("Failed conversion to EC_Scalar_Data_PC");
+   }
+   return *p;
+}
+
+const std::shared_ptr<const EC_Group_Data>& EC_Scalar_Data_PC::group() const {
+   return m_group;
+}
+
+size_t EC_Scalar_Data_PC::bytes() const {
+   return this->group()->order_bytes();
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::clone() const {
+   return std::make_unique<EC_Scalar_Data_PC>(this->group(), this->value());
+}
+
+bool EC_Scalar_Data_PC::is_zero() const {
+   return this->value().is_zero();
+}
+
+bool EC_Scalar_Data_PC::is_eq(const EC_Scalar_Data& other) const {
+   return (value() == checked_ref(other).value());
+}
+
+void EC_Scalar_Data_PC::assign(const EC_Scalar_Data& other) {
+   m_v = checked_ref(other).value();
+}
+
+void EC_Scalar_Data_PC::square_self() {
+   // TODO square in place
+   m_v = m_v.square();
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::negate() const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v.negate());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::invert() const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v.invert());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::add(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v + checked_ref(other).value());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::sub(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v - checked_ref(other).value());
+}
+
+std::unique_ptr<EC_Scalar_Data> EC_Scalar_Data_PC::mul(const EC_Scalar_Data& other) const {
+   return std::make_unique<EC_Scalar_Data_PC>(m_group, m_v * checked_ref(other).value());
+}
+
+void EC_Scalar_Data_PC::serialize_to(std::span<uint8_t> bytes) const {
+   BOTAN_ARG_CHECK(bytes.size() == m_group->order_bytes(), "Invalid output length");
+   m_group->pcurve().serialize_scalar(bytes, m_v);
+}
+
+EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
+                                               PCurve::PrimeOrderCurve::AffinePoint pt) :
+      m_group(std::move(group)), m_pt(std::move(pt)) {
+   m_bytes = m_pt.serialize<secure_vector<uint8_t>>();
+}
+
+EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
+                                               std::span<const uint8_t> bytes) :
+      m_group(std::move(group)), m_pt(deserialize_pcurve_pt(m_group->pcurve(), bytes)) {
+   m_bytes = m_pt.serialize<secure_vector<uint8_t>>();
+}
```

Comment:
Two things:

1. `m_bytes` could be initialized in the initializer-list as well

```suggestion
EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
                                               PCurve::PrimeOrderCurve::AffinePoint pt) :
      m_group(std::move(group)), m_pt(std::move(pt)), m_bytes(m_pt.serialize<secure_vector<uint8_t>>()) {
}

EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
                                               std::span<const uint8_t> bytes) :
      m_group(std::move(group)), m_pt(deserialize_pcurve_pt(m_group->pcurve(), bytes)), m_bytes(m_pt.serialize<secure_vector<uint8_t>>()) {
}
```

2. Given that `m_bytes` is never written to, perhaps its worthwhile to add `std::span<>`s that centrally refer to the segments of the serialization. That would avoid spreading the knowledge about the structure in `m_bytes` across several methods. Like so:

```C++
EC_AffinePoint_Data_PC::EC_AffinePoint_Data_PC(std::shared_ptr<const EC_Group_Data> group,
                                               PCurve::PrimeOrderCurve::AffinePoint pt) :
      m_group(std::move(group)), m_pt(std::move(pt)), m_bytes(m_pt.serialize<secure_vector<uint8_t>>()),
      m_bytes_x(std::span{m_bytes}.subspan(1, field_element_bytes())),
      m_bytes_y(std::span{m_bytes}.last(field_element_bytes())) {
}
```

Reply:
Both done